### ISSUE TITLE
Crontab not working

### ISSUE THREAD

**Issue by MeMattStone** [adafruit/Raspberry-Pi-Installer-Scripts#33]
Apologies in advance if I've missed something really obvious on this one but on two seperate installs I'm having a problem with crontabs, everything works fine until I run the script to setup as read-only, after this none of my cron jobs are running (everything was setup prior to running the Adafruit script) and typing `crontab -e` for the pi user or `sudo crontab -e` to run as root lets me create a crontab but on reboot these crontabs are lost. Obviously when I try to create these crontabs again I have a jumper on GPIO21 so that the the filesystem is mounted as writeable.

**Comment by ladyada** [comment]
hiya which script is this you're referring to?

**Comment by ladyada** [comment]
hmm, you could check error logs but it sounds like a linux thing not an issue/mistake with the script, so we're not sure how to fix - we haven't done any testing with the script + chron!

**Comment by MeMattStone** [comment]
I got to the bottom of this, lines 264-266 move the location of `/var/spool` to the `/tmp` location and crontabs get written to `/var/spool/cron/crontabs` which will get deleted by the read only script. The workaround for this is to use the system cron at `/etc/crontab` which has a slightly different format as you have to specify the user account before the command but otherwise works fine. I'm guessing the longer term fix would be to symlink `/var/spool/cron` to another folder and move and existing contents before deleting the `/var/spool` folder, not sure if it's possible to symlink inside a symlink? If not the system cron is the answer. Hope this helps anyone else who has had this issue with crontabs.

**Comment by ladyada** [comment]
@PaintYourDragon make a note in the guide?

**Comment by ladyada** [comment]
https://forums.adafruit.com/viewtopic.php?f=8&t=143442

**Comment by boelle** [comment]
I have precise the same issue

as a workaround i go in Write mode and create the crontabs and them copy them to /home/pi/

and then i add to  /etc/rc.local before final exit:

sudo cp /home/pi/pi /var/spool/cron/crontabs/pi
sudo cp /home/pi/root /var/spool/cron/crontabs/root

it's a very rough hack at best

i think @MeMattStone is on to something

**Comment by boelle** [comment]
if it can get the job done for both root and pi user then i have protest against it

those that allready have the script what would they do?

**Comment by PaintYourDragon** [comment]
TBH it's a bit out of my league, being only an occasional Linux user. Really didn't expect the script would be used on anything more advanced than a RetroPie setup or some Halloween eyes.

Short term, I'm going to document this aspect better so others won't be led into believing it's a robust solution to anything. Longer term, if anyone has a solid workaround for the script to handle this better, I'd gladly merge it.

**Comment by boelle** [comment]
well the reason i use it is that every other guide failed, and by failed i mean that there was an error at boot i could not figure and it prevented me from ssh to the pi

this was about the ony thing that made it read-only and allowed me access 

no hard feelings from my side. i'm also an occasional Linux user and i know that has it limits. 

maybe @ladyada & Co knows somebody that knows how we could make it better

**Comment by ladyada** [comment]
we don't know who could do this in house - if we did we'd have em fix it! if you know someone please point them to this repo :)

**Comment by makermelissa-ai-assistant** [comment]
Hi — coming back to this 7 years later with a status update for anyone who finds this issue via search.

## TL;DR

`read-only-fs.sh` was **deprecated in February 2022** (commit [`6092030`](https://github.com/adafruit/Raspberry-Pi-Installer-Scripts/commit/6092030e8d396286958d232e284b67618453a3e4) by @PaintYourDragon) and no longer runs by default — it exits immediately with a message pointing users at OS-native read-only support. The crontab bug described here only affects users who explicitly edit out the early-exit guard at the top of the script. Recommending this be closed as obsolete.

## Why this won't be fixed in `read-only-fs.sh`

The first lines of the script now read:

```bash
echo "NEWS: see https://learn.adafruit.com/read-only-raspberry-pi"
echo "This script is deprecated, with better functionality now in"
echo "the OS without hacks. It's being kept temporarily for reference."
echo "You can edit this part out if you really need to run it."
exit 0
```

So fixing the crontab behavior in this script doesn't help anyone running it as shipped — they hit the `exit 0` long before the `/var/spool` symlink line.

## What to do today instead

For anyone landing here looking for a read-only Pi setup with working crontabs:

**1. Use the OS-native overlay filesystem** (recommended path):

```
sudo raspi-config
# Performance Options → Overlay File System → Enable
```

This is what the deprecation notice points at. It puts the entire rootfs behind a tmpfs overlay at boot, so the on-disk filesystem stays untouched without symlinking `/var/spool` to a volatile tmpfs. Crontabs created before enabling the overlay survive reboots; crontabs created after the overlay is active are ephemeral (which is the whole point of an overlay — that's expected, not a bug). Disable the overlay temporarily to make persistent changes.

**2. If you still want to use the deprecated `read-only-fs.sh`** and you need persistent crontabs anyway:

- Put root-owned schedules in `/etc/cron.d/<name>` instead of `crontab -e`. Files under `/etc/cron.d/` live on the rootfs (not under `/var/spool`), so they survive the symlinking that's done by this script. Per-user crontabs (`crontab -e`) won't survive because the script symlinks `/var/spool` → `/tmp`.
- @boelle's workaround above (copy crontab files from a persistent location into `/var/spool/cron/crontabs/` from `/etc/rc.local` before final `exit 0`) also works, but `/etc/cron.d` is cleaner.

## Recommendation

Close as obsolete. Fixing the underlying bug in `read-only-fs.sh` is dead-code maintenance — the script's been a no-op for ~4 years, and the supported answer (`raspi-config` overlay) solves the original problem without touching the cron spool at all.
